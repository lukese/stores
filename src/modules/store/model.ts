import { getModelForClass } from "@typegoose/typegoose";
import { ObjectId } from "mongodb";

import { Store } from "../../entities/store";

export const StoreMongooseModel = getModelForClass(Store);

export default class StoreModel {
  async getById(_id: ObjectId): Promise<Store | null> {
    return StoreMongooseModel.findById(_id).lean().exec();
  }

  async getNearestLocations(latitude: number, longitude: number): Promise<Store[] | null> {
    return StoreMongooseModel.find(
        { location :
              { $near :
                    {
                      $geometry : {
                        type : "Point" ,
                        coordinates : [latitude, longitude] },
                      $maxDistance : 10000000
                    }
              }
        }
    ).limit(5).exec();
  }
}
