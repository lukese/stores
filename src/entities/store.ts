import { ObjectId } from "mongodb";
import { ObjectType, Field } from "type-graphql";
import { prop } from "@typegoose/typegoose";
import { ID, Float } from "type-graphql";
import 'reflect-metadata';

@ObjectType()
export class Store {
    @Field(type => ID)
    readonly _id!: ObjectId;

    @Field()
    city: string;

    @Field()
    postalCode: string;

    @prop()
    @Field()
    street: string;

    @prop()
    @Field()
    street2: string;

    @prop()
    @Field()
    street3: string;

    @prop()
    @Field()
    addressName: string;

    @prop()
    @Field()
    uuid: string;

    @prop()
    @Field(type => Float)
    longitude: number;

    @prop()
    @Field(type => Float)
    latitude: number;

    @prop()
    @Field()
    complexNumber: string;

    @prop()
    @Field()
    showWarningMessage: boolean;

    @prop()
    @Field()
    todayOpen: string;

    @prop()
    @Field()
    locationType: string;

    @prop()
    @Field()
    collectionPoint: boolean;

    @prop()
    @Field()
    sapStoreID: string;

    @prop()
    @Field()
    todayClose: string;
}
