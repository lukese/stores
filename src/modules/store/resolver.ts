import { Resolver, Arg, Query } from "type-graphql";
import { Service } from "typedi";
import { ObjectId } from "mongodb";

import { Store } from "../../entities/store";
import StoreService from "./service";

@Service()
@Resolver((of) => Store)
export default class StoreResolver {
  constructor(private readonly storeService: StoreService) {}

  @Query((returns) => Store)
  async getStore(@Arg("id") id: ObjectId) {
    return this.storeService.getById(id);
  }

  @Query((returns) => [Store])
  async getNearestLocations(@Arg("latitude") latitude: number, @Arg("longitude") longitude: number) {
    return await this.storeService.getNearestLocations(latitude, longitude);
  }
}
