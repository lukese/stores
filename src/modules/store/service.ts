import { Service } from "typedi";
import { ObjectId } from "mongodb";

import StoreModel from "./model";
import { Store } from "../../entities/store";

@Service()
export default class StoreService {
  constructor(private readonly storeModel: StoreModel) {}

  public async getById(_id: ObjectId): Promise<Store | null> {
    const store = await this.storeModel.getById(_id);

    return store;
  }
}
