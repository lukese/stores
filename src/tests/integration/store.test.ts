import { createTestClient } from "apollo-server-testing";
import { ApolloServer, gql } from "apollo-server-express";
import { buildSchema } from "../../utils";
// @ts-ignore
import mongoose from "mongoose";

import { StoreMongooseModel } from "../../modules/store/model";

import {
    connect,
    clearDatabase,
    closeDatabase,
    populateDatabase,
} from "../utils";

beforeAll(async () => connect());

beforeEach(async () => {
    await populateDatabase(StoreMongooseModel, []);
});

afterEach(async () => {
    await clearDatabase();
});

afterAll(async (done) => {
    await closeDatabase();
    done();
});

describe("Store", () => {
	it(`should get a store`, async () => {
      const storeId = new mongoose.Types.ObjectId().toHexString().toString();

      await populateDatabase(StoreMongooseModel, [
          {
              _id: storeId,
              street: 'Test',
              longitude: 42.3223,
              latitude: 42.23234,
          },
      ]);

      const graphQLSchema = await buildSchema();

      const server = new ApolloServer({
          schema: graphQLSchema,
      }) as any;

      const { query } = createTestClient(server);

      const variables = {
          id: storeId,
      };

		const GET_STORE = gql`
        query getStore($id: ObjectId!) {
            getStore(id: $id) {
                _id
                longitude
                latitude
            }
        }
    `;

      const res = await query({
          query: GET_STORE,
          variables,
      });

      // @ts-ignore
        expect(res).toMatchSnapshot({
          'data': {
              'getStore': {
                  _id: expect.any(String),
              }
          }
      });
	});
});
