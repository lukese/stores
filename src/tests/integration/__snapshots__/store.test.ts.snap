// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store should get a store 1`] = `
Object {
  "data": Object {
    "getStore": Object {
      "_id": Any<String>,
      "latitude": 42.23234,
      "longitude": 42.3223,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
